File: backend-go/internal/handlers/accounts.go
package handlers

import (
	"net/http"

	"github.com/chuyen-gia-crypto/backend/pkg/logger"
	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

// GetAccounts handles GET /api/accounts
func GetAccounts(c *gin.Context) {
	logger.Info("GET /api/accounts called")

	if Deps == nil || Deps.AccountRepo == nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{
			"success": false,
			"error":   "Database not initialized",
		})
		return
	}

	accounts, err := Deps.AccountRepo.GetAll(c.Request.Context())
	if err != nil {
		logger.Error("Failed to get accounts", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to retrieve accounts",
		})
		return
	}

	// Convert to DTO with proper time formatting
	result := make([]gin.H, len(accounts))
	for i, acc := range accounts {
		result[i] = gin.H{
			"id":                 acc.ID,
			"symbol":             acc.Symbol,
			"method_id":          acc.MethodID,
			"starting_balance":   acc.StartingBalance,
			"current_balance":    acc.CurrentBalance,
			"equity":             acc.Equity,
			"unrealized_pnl":     acc.UnrealizedPnl,
			"realized_pnl":       acc.RealizedPnl,
			"total_trades":       acc.TotalTrades,
			"winning_trades":     acc.WinningTrades,
			"losing_trades":      acc.LosingTrades,
			"max_drawdown":       acc.MaxDrawdown,
			"consecutive_losses": acc.ConsecutiveLosses,
			"last_trade_time":    formatTimePtr(acc.LastTradeTime),
			"cooldown_until":     formatTimePtr(acc.CooldownUntil),
			"created_at":         formatTime(acc.CreatedAt),
			"updated_at":         formatTime(acc.UpdatedAt),
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"data":    result,
	})
}

// ResetAccount handles POST /api/accounts/reset
func ResetAccount(c *gin.Context) {
	logger.Info("POST /api/accounts/reset called")

	symbol := c.Query("symbol")
	methodID := c.Query("method_id")

	if symbol == "" || methodID == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "symbol and method_id are required",
		})
		return
	}

	if Deps == nil || Deps.AccountRepo == nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{
			"success": false,
			"error":   "Database not initialized",
		})
		return
	}

	account, err := Deps.AccountRepo.Reset(c.Request.Context(), symbol, methodID)
	if err != nil {
		logger.Error("Failed to reset account", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to reset account",
		})
		return
	}

	// Convert to DTO with proper time formatting
	result := gin.H{
		"id":                 account.ID,
		"symbol":             account.Symbol,
		"method_id":          account.MethodID,
		"starting_balance":   account.StartingBalance,
		"current_balance":    account.CurrentBalance,
		"equity":             account.Equity,
		"unrealized_pnl":     account.UnrealizedPnl,
		"realized_pnl":       account.RealizedPnl,
		"total_trades":       account.TotalTrades,
		"winning_trades":     account.WinningTrades,
		"losing_trades":      account.LosingTrades,
		"max_drawdown":       account.MaxDrawdown,
		"consecutive_losses": account.ConsecutiveLosses,
		"last_trade_time":    formatTimePtr(account.LastTradeTime),
		"cooldown_until":     formatTimePtr(account.CooldownUntil),
		"created_at":         formatTime(account.CreatedAt),
		"updated_at":         formatTime(account.UpdatedAt),
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"data":    result,
	})
}

File: backend/src/routes/accounts.ts
import { Router, Request, Response } from 'express';
import { prisma } from '../lib/prisma';

const router = Router();

// Extend Express Request to include prisma
declare global {
  namespace Express {
    interface Request {
      prisma?: typeof prisma;
    }
  }
}

// GET /api/accounts - Get all accounts (or filter by method)
router.get('/', async (req: Request, res: Response) => {
  const prismaClient = req.prisma || prisma;

  if (!prismaClient) {
    return res.status(503).json({
      success: false,
      error: 'Database not available'
    });
  }

  const { method } = req.query;

  try {
    let accounts;
    if (method) {
      accounts = await prismaClient.account.findMany({
        where: { method_id: String(method) },
        orderBy: { created_at: 'desc' }
      });
    } else {
      accounts = await prismaClient.account.findMany({
        orderBy: { created_at: 'desc' }
      });
    }
    
    return res.json({
      success: true,
      data: accounts,
      meta: { count: accounts.length, method: method || null }
    });
  } catch (error: any) {
    return res.status(500).json({
      success: false,
      error: error.message
    });
  }
});

// GET /api/accounts/:symbol - Get account by symbol (and optionally method)
router.get('/:symbol', async (req: Request, res: Response) => {
  const prismaClient = req.prisma || prisma;

  if (!prismaClient) {
    return res.status(503).json({
      success: false,
      error: 'Database not available'
    });
  }

  const { symbol } = req.params;
  const { method } = req.query;

  try {
    let account;
    if (method) {
      account = await prismaClient.account.findUnique({
        where: { symbol_method_id: { symbol: String(symbol).toUpperCase(), method_id: String(method) } }
      });
    } else {
      account = await prismaClient.account.findFirst({
        where: { symbol: String(symbol).toUpperCase() }
      });
    }
    
    if (!account) {
      return res.status(404).json({
        success: false,
        error: 'Account not found'
      });
    }
    
    return res.json({
      success: true,
      data: account,
      meta: { method: method || null }
    });
  } catch (error: any) {
    return res.status(500).json({
      success: false,
      error: error.message
    });
  }
});

// POST /api/accounts/reset/:symbol - Reset account to starting balance (and optionally method)
router.post('/reset/:symbol', async (req: Request, res: Response) => {
  const prismaClient = req.prisma || prisma;

  if (!prismaClient) {
    return res.status(503).json({
      success: false,
      error: 'Database not available'
    });
  }

  const { symbol } = req.params;
  const { method = 'ict' } = req.body;

  try {
    const account = await prismaClient.account.findUnique({
      where: { symbol_method_id: { symbol: String(symbol).toUpperCase(), method_id: String(method) } }
    });
    
    if (!account) {
      return res.status(404).json({
        success: false,
        error: 'Account not found'
      });
    }
    
    const updatedAccount = await prismaClient.account.update({
      where: { symbol_method_id: { symbol: String(symbol).toUpperCase(), method_id: String(method) } },
      data: {
        current_balance: account.starting_balance,
        equity: account.starting_balance,
        unrealized_pnl: 0,
        realized_pnl: 0,
        total_trades: 0,
        winning_trades: 0,
        losing_trades: 0,
        max_drawdown: 0,
        consecutive_losses: 0,
        last_trade_time: null,
        cooldown_until: null
      }
    });
    
    return res.json({
      success: true,
      data: updatedAccount,
      meta: { method },
      message: 'Account reset successfully'
    });
  } catch (error: any) {
    return res.status(500).json({
      success: false,
      error: error.message
    });
  }
});

export default router;
